File: leetCode/tree/Problem098.java
package leetCode.tree;

import java.util.Stack;

/**
 * Given a binary tree, determine if it is a valid binary search tree (BST).
 * Assume a BST is defined as follows:
 * The left subtree of a node contains only nodes with keys less than the node's key.
 * The right subtree of a node contains only nodes with keys greater than the node's key.
 * Both the left and right subtrees must also be binary search trees.
 */
public class Problem098 {
    private class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }

    // recursive solution
    public boolean isValidBST(TreeNode root) {
        if (root == null) return true;
        if (!isValidBST(root.left)) return false;
        if (!isValidBST(root.right)) return false;

        TreeNode temp = root.left;
        if (temp != null) {
            while (temp.right != null) temp = temp.right;
            if (temp.val >= root.val) return false;
        }

        temp = root.right;
        if (temp != null) {
            while (temp.left != null) temp = temp.left;
            if (temp.val <= root.val) return false;
        }
        return true;
    }

    // inorder traversal, check whether it is an incresing sequences
    public boolean isValidBST1(TreeNode root) {
        Stack<TreeNode> stack = new Stack<>();
        TreeNode temp = root, pre = null;

        while (temp != null || !stack.isEmpty()) {
            while (temp != null) {
                stack.push(temp);
                temp = temp.left;
            }

            if (!stack.isEmpty()) {
                temp = stack.pop();
                if (pre != null && pre.val >= temp.val) return false;
                pre = temp;
                temp = temp.right;
            }
        }
        return true;
    }
}

File: leetCode/tree/Problem098.cpp
#include <climits>
#include <stack>

#include "tree.h"

// recursive solution
bool isvalid(TreeNode*, int, int, bool, bool);
bool isValidBST(TreeNode* root) {
  return isvalid(root, INT_MIN, INT_MAX, true, true);
}

bool isvalid(TreeNode* root, int lo, int hi, bool inclu_lo, bool inclu_hi) {
  if (!root) return true;
  if ((inclu_lo && root->val < lo) || (!inclu_lo && root->val <= lo) ||
      (inclu_hi && root->val > hi) || (!inclu_hi && root->val >= hi))
    return false;
  return isvalid(root->left, lo, root->val, inclu_lo, false) &&
         isvalid(root->right, root->val, hi, false, inclu_hi);
}

// iterative solution
bool isValidBST2(TreeNode* root) {
  std::stack<TreeNode*> st;
  TreeNode *cur = root, *prev = nullptr;
  while (!st.empty() || cur) {
    while (cur) {
      st.push(cur);
      cur = cur->left;
    }
    if (!st.empty()) {
      cur = st.top();
      st.pop();
      if (prev && cur->val <= prev->val) return false;
      prev = cur;
      cur = cur->right;
    }
  }
  return true;
}